flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/AbstractStateTableSnapshot.java

```diff
@@ -19,34 +19,106 @@
 package org.apache.flink.runtime.state.heap;
 
 import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataOutputView;
 import org.apache.flink.runtime.state.StateSnapshot;
+import org.apache.flink.runtime.state.StateSnapshotTransformer;
+import org.apache.flink.runtime.state.metainfo.StateMetaInfoSnapshot;
 import org.apache.flink.util.Preconditions;
 
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
 /**
  * Abstract base class for snapshots of a {@link StateTable}. Offers a way to serialize the snapshot (by key-group).
  * All snapshots should be released after usage.
  */
 @Internal
-abstract class AbstractStateTableSnapshot<K, N, S, T extends StateTable<K, N, S>> implements StateSnapshot {
+abstract class AbstractStateTableSnapshot<K, N, S, T extends StateTable<K, N, S>>
```

Comment:
I think there is no need to introduce `getStateMapSnapshotArray` for `NestedStateTable`. For `CopyOnWriteStateTable`, we should create snapshots for all `StateMap`s in the sync part of the snapshot for backend, hold them somewhere like in an array, and use them to write state to output in the async part. But for `NestedStateTable` which only supports synchronous snapshot, we can create the snapshot of `StateMap` and write state to output one by one, and there is no need to hold all `StateMap`'s snapshots firstly and then write state to output

Reply:
I can see your point. One could implement the structure of the snapshots in a way that is eager for copy-on-write and lazy for the nested table, so that they can still have the same interface, but I would not make that a blocker for the PR and think the current solution is also ok.